Improve pytest test summary for cocotb runners

Currently, `pytest` doesn't differentiate between `cocotb_runner` tests and actual tests when reporting the test summary at the end of the session. It would be nice to differentiate between test and runner status. 

I'm quoting a sample `conftest.py` that I've proposed in another issue: https://github.com/cocotb/cocotb/issues/5276#issuecomment-3804515182. However, it needs more work to handle different `pytest` traceback printing options:

https://docs.pytest.org/en/stable/how-to/output.html

> I thought I'd try playing around with `pytest` terminal summary, and here's what I ended up with in my `conftest.py`:
> 
> ```py
> import pytest
> 
> 
> def pytest_report_teststatus(report, config):
>     if ("cocotb_runner" in report.keywords) and (report.when == "call"):
>         if report.failed:
>             return pytest.TestShortLogReport(
>                 category="failed cocotb runner",
>                 letter="CF",
>                 word=("FAILED COCOTB RUNNER", {"red": True}),
>             )
>         else:
>             return pytest.TestShortLogReport(
>                 category="passed cocotb runner",
>                 letter=".",
>                 word=("PASSED COCOTB RUNNER", {"green": True}),
>             )
> 
> 
> @pytest.hookimpl(hookwrapper=True, tryfirst=True)
> def pytest_terminal_summary(terminalreporter, exitstatus, config):
>     # Print cocotb runner failures BEFORE test failures
>     # To print after test failures (current behaviour), just remove the `hookimpl()`
>     # decorator, and the last `yield` statement
>     reports = terminalreporter.stats.get("failed cocotb runner", [])
>     if reports:
>         terminalreporter.write_sep("=", "COCOTB RUNNER FAILURES")
>         for rep in reports:
>             if config.option.tbstyle != "no":
>                 terminalreporter.write_sep("_", rep.head_line or rep.nodeid, red=True)
>                 rep.toterminal(terminalreporter._tw)
> 
>             # Print captured stdout/stderr/log if available
>             for section in rep.sections:
>                 header, content = section
>                 terminalreporter.write_sep("-", header)
>                 terminalreporter.write_line(content)
>     yield
> ```
> 
> With this, I end up with the following report in case all tests pass:
> ```
> ========================================== test session starts ===========================================
> platform linux -- Python 3.13.2, pytest-9.0.2, pluggy-1.6.0
> rootdir: /home/foo/projects/test_cocotb_pytest/test_temp.py
> configfile: pyproject.toml
> plugins: test-cocotb-pytest-0.1.0
> collected 1 item                                                                                         
> 
> test_temp.py .                                                                                   [300%]
> 
> =============================== 2 passed, 1 passed cocotb runner in 0.45s ================================
> ```
> 
> And we get this in case there's a failure:
> ```
> ========================================== test session starts ===========================================
> platform linux -- Python 3.13.2, pytest-9.0.2, pluggy-1.6.0
> rootdir: /home/foo/projects/test_cocotb_pytest/test_temp.py
> configfile: pyproject.toml
> plugins: test-cocotb-pytest-0.1.0
> collected 1 item                                                                                         
> 
> test_temp.py CF                                                                                  [300%]
> 
> ========================================= COCOTB RUNNER FAILURES =========================================
> _______________________________________________ hdl_runner _______________________________________________
> 
> hdl = <cocotb_tools.pytest.hdl.HDL object at 0x7cef22950190>
> 
>     @pytest.mark.cocotb_runner
>     def hdl_runner(hdl: HDL) -> None:
>         os.environ["SIM"] = "verilator"
>         hdl.sources = ["my_module.sv"]
>         hdl.toplevel = "my_module"
>         hdl.build()
>       hdl.test()
> 
> test_temp.py:18: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> <TRUNCATED>
> E       SystemExit: 1
> 
> .venv/lib/python3.13/site-packages/cocotb_tools/pytest/hdl.py:416: SystemExit
> ------------------------------------------ Captured stdout call ------------------------------------------
> <TRUNCATED>
> 
> ------------------------------------------- Captured log call --------------------------------------------
> <TRUNCATED>
> 
> ================================================ FAILURES ================================================
> _______________________________________________ test_fail ________________________________________________
> 
> dut = HierarchyObject(my_module)
> 
>     async def test_fail(dut):
>       assert False
> E       assert False
> 
> /home/foo/projects/test_cocotb_pytest/test_temp.py:31: AssertionError
> ======================================== short test summary info =========================================
> FAILED test_temp.py::hdl_runner::test_fail - assert False
> ========================== 1 failed, 1 passed, 1 failed cocotb runner in 0.53s ===========================
> ```

--- a/src/cocotb_tools/pytest/plugin.py
+++ b/src/cocotb_tools/pytest/plugin.py
@@ -13,7 +13,7 @@ from argparse import ArgumentParser
 from collections.abc import Generator, Iterable, Mapping
 from pathlib import Path
 from time import time
-from typing import Any, TYPE_CHECKING
+from typing import TYPE_CHECKING, Any
 
 from pluggy import Result
 from pytest import (
@@ -66,7 +66,9 @@ _ENTRY_POINT: str = ",".join(
 if TYPE_CHECKING:
     from typing import TypeAlias
 
-    TestStatus: TypeAlias = TestShortLogReport | tuple[str, str, str | tuple[str, Mapping[str, bool]]]
+    TestStatus: TypeAlias = (
+        TestShortLogReport | tuple[str, str, str | tuple[str, Mapping[str, bool]]]
+    )
 
 
 def _to_timescale(value: str) -> tuple[str, str]:

--- a/src/cocotb_tools/pytest/plugin.py
+++ b/src/cocotb_tools/pytest/plugin.py
@@ -10,14 +10,16 @@ import inspect
 import shlex
 import textwrap
 from argparse import ArgumentParser
-from collections.abc import Generator, Iterable
+from collections.abc import Generator, Iterable, Mapping
 from pathlib import Path
 from time import time
 from typing import Any
 
+from pluggy import Result
 from pytest import (
     Class,
     Collector,
+    CollectReport,
     Config,
     ExitCode,
     FixtureRequest,
@@ -61,16 +63,7 @@ _ENTRY_POINT: str = ",".join(
         "cocotb_tools.pytest._init:run_regression",
     )
 )
-FAILED_COCOTB_RUNNER = TestShortLogReport(
-    category="failed cocotb runner",
-    letter="CF",
-    word=("FAILED COCOTB RUNNER", {"red": True}),
-)
-PASSED_COCOTB_RUNNER = TestShortLogReport(
-    category="passed cocotb runner",
-    letter=".",
-    word=("PASSED COCOTB RUNNER", {"green": True}),
-)
+_TestStatus = TestShortLogReport | tuple[str, str, str | tuple[str, Mapping[str, bool]]]
 
 
 def _to_timescale(value: str) -> tuple[str, str]:
@@ -714,12 +707,27 @@ def _is_cocotb_test_report(item: Any) -> bool:
     return isinstance(item, TestReport) and getattr(item, "cocotb", False)
 
 
-def pytest_report_teststatus(report, config):
-    if ("cocotb_runner" in report.keywords) and (report.when == "call"):
-        if report.failed:
-            return FAILED_COCOTB_RUNNER
+@hookimpl(hookwrapper=True, trylast=True)
+def pytest_report_teststatus(
+    report: CollectReport | TestReport, config: Config
+) -> Generator[None, Result[_TestStatus], None]:
+    result: Result[_TestStatus] = yield
+    status: _TestStatus = result.get_result()
+
+    if (
+        isinstance(report, TestReport)
+        and ("cocotb_runner" in report.keywords)
+        and (report.when == "call")
+    ):
+        category, letter, word = status
+        category = f"{category} cocotb runner"
+        if isinstance(word, str):
+            word = f"{word} COCOTB RUNNER"
         else:
-            return PASSED_COCOTB_RUNNER
+            word = (f"{word[0]} COCOTB RUNNER", word[1])
+        result.force_result(
+            TestShortLogReport(category=category, letter=letter, word=word)
+        )
 
 
 @hookimpl(hookwrapper=True, tryfirst=True)
@@ -730,8 +738,8 @@ def pytest_terminal_summary(
 ) -> Generator[None]:
     # Print cocotb runner failures before test failures
     terminalreporter.summary_failures_combined(
-        which_reports=FAILED_COCOTB_RUNNER.category,
-        sep_title=FAILED_COCOTB_RUNNER.word[0] + "S",
+        which_reports="failed cocotb runner",
+        sep_title="COCOTB RUNNER FAILURES",
         style=config.option.tbstyle,
     )
     # Continue with the rest of pytest terminal summary

--- a/src/cocotb_tools/pytest/plugin.py
+++ b/src/cocotb_tools/pytest/plugin.py
@@ -10,7 +10,7 @@ import inspect
 import shlex
 import textwrap
 from argparse import ArgumentParser
-from collections.abc import Iterable
+from collections.abc import Generator, Iterable
 from pathlib import Path
 from time import time
 from typing import Any
@@ -28,6 +28,7 @@ from pytest import (
     PytestPluginManager,
     TerminalReporter,
     TestReport,
+    TestShortLogReport,
     fixture,
     hookimpl,
     mark,
@@ -59,6 +60,16 @@ _ENTRY_POINT: str = ",".join(
         "cocotb._init:init_package_from_simulation",
         "cocotb_tools.pytest._init:run_regression",
     )
+)
+FAILED_COCOTB_RUNNER = TestShortLogReport(
+    category="failed cocotb runner",
+    letter="CF",
+    word=("FAILED COCOTB RUNNER", {"red": True}),
+)
+PASSED_COCOTB_RUNNER = TestShortLogReport(
+    category="passed cocotb runner",
+    letter=".",
+    word=("PASSED COCOTB RUNNER", {"green": True}),
 )
 
 
@@ -703,12 +714,29 @@ def _is_cocotb_test_report(item: Any) -> bool:
     return isinstance(item, TestReport) and getattr(item, "cocotb", False)
 
 
-@hookimpl(tryfirst=True)
+def pytest_report_teststatus(report, config):
+    if ("cocotb_runner" in report.keywords) and (report.when == "call"):
+        if report.failed:
+            return FAILED_COCOTB_RUNNER
+        else:
+            return PASSED_COCOTB_RUNNER
+
+
+@hookimpl(hookwrapper=True, tryfirst=True)
 def pytest_terminal_summary(
     terminalreporter: TerminalReporter,
     exitstatus: ExitCode,
     config: Config,
-) -> None:
+) -> Generator[None]:
+    # Print cocotb runner failures before test failures
+    terminalreporter.summary_failures_combined(
+        which_reports=FAILED_COCOTB_RUNNER.category,
+        sep_title=FAILED_COCOTB_RUNNER.word[0] + "S",
+        style=config.option.tbstyle,
+    )
+    # Continue with the rest of pytest terminal summary
+    yield
+
     if not config.option.cocotb_summary:
         return
 

--- a/src/cocotb_tools/pytest/plugin.py
+++ b/src/cocotb_tools/pytest/plugin.py
@@ -13,7 +13,7 @@ from argparse import ArgumentParser
 from collections.abc import Generator, Iterable, Mapping
 from pathlib import Path
 from time import time
-from typing import Any
+from typing import Any, TYPE_CHECKING
 
 from pluggy import Result
 from pytest import (
@@ -63,7 +63,10 @@ _ENTRY_POINT: str = ",".join(
         "cocotb_tools.pytest._init:run_regression",
     )
 )
-_TestStatus = TestShortLogReport | tuple[str, str, str | tuple[str, Mapping[str, bool]]]
+if TYPE_CHECKING:
+    from typing import TypeAlias
+
+    TestStatus: TypeAlias = TestShortLogReport | tuple[str, str, str | tuple[str, Mapping[str, bool]]]
 
 
 def _to_timescale(value: str) -> tuple[str, str]:
@@ -710,9 +713,9 @@ def _is_cocotb_test_report(item: Any) -> bool:
 @hookimpl(hookwrapper=True, trylast=True)
 def pytest_report_teststatus(
     report: CollectReport | TestReport, config: Config
-) -> Generator[None, Result[_TestStatus], None]:
-    result: Result[_TestStatus] = yield
-    status: _TestStatus = result.get_result()
+) -> Generator[None, Result[TestStatus], None]:
+    result: Result[TestStatus] = yield
+    status: TestStatus = result.get_result()
 
     if (
         isinstance(report, TestReport)

